### DISCUSSION TITLE
python3 update that is not wanted

### DISCUSSION THREAD

**Discussion by jraman** [Homebrew/discussions#974] [Everyday usage]
When I installed `rsync` for the first time, it updated the python3.9 install (from 3.9.1 to 3.9.2_1).  I guess that's OK.

But it caused the following problems--
 * It wiped out all the packages installed in 3.9.1
 * It broke pip itself:
 ```
$ pip3
Traceback (most recent call last):
  File "/usr/local/bin/pip3", line 5, in <module>
    from pip._internal.cli.main import main
ModuleNotFoundError: No module named 'pip'
```

This is quite annoying.  Why does it do the update?  Can it be overridden?  If not, how can the update be done smoothly so that the installed packages are carried over?

Thanks.

**Comment by Bo98** [comment]
This sounds like `brew postinstall python@3.9` failed to complete. Try run it again - do you see any errors output?

**Reply by jraman** [reply]
I ended up running `brew reinstall python3` based on a google search and now I'm not sure if that was the right thing to do -- it has been running for a few hours now (pulling in gcc, rust, imagemagick, to name a few).  It is still running.  I hope this ends well for me.

I don't see why python got updated from 3.9.1 to 3.9.2.  Does rsync-3.2.3 really require 3.9.2 as opposed to 3.9.1?

**Reply by jonchang** [reply]
Homebrew doesn’t support arbitrary mixing and matching of formula versions, so everything a formula depends on, and everything that depends on it in turn, needs to be upgraded to the latest version as that’s the only combination of formulae we test. As a consequence any given upgrade or install command can upgrade many other (seemingly unrelated) formulae, if something important like python or openssl also needed an upgrade.

https://docs.brew.sh/FAQ#why-does-brew-upgrade-formula-also-upgrade-a-bunch-of-other-stuff

**Reply by SMillerDev** [reply]
> I ended up running brew reinstall python3 based on a google search and now I'm not sure if that was the right thing to do -- it has been running for a few hours now (pulling in gcc, rust, imagemagick, to name a few). It is still running. I hope this ends well for me.

This probably means you don't have command line tools installed. What does `brew config` say?

**Reply by jraman** [reply]
@SMillerDev, you might be right.  This is what I get with `brew config`:
```
Xcode: N/A
```

Is this what you meant?

Also, the install completed -- `rust` failed to build (I think I can live without it for now) and a few failed symlinks which I should be able to fix.

**Reply by jraman** [reply]
@jonchang, I understand that arbitrary mixing is a no-no.  But a minor version?  I'm familiar with a `pip`-like specification like `python >= 3.9` which would pass with 3.9.1.  

I would understand 3.7 needing to get upgraded to 3.9 if that's what was required, but not sure about rsync (probably one of its dependencies) requiring 3.9.2 and not being satisfied with 3.9.1.

**Reply by SMillerDev** [reply]
Yes, python requires the CLT to use pre-built binaries. You can run `xcode-select --install` and after that `brew reinstall python` should only take a couple of minutes.

**Reply by SMillerDev** [reply]
> I understand that arbitrary mixing is a no-no. But a minor version? I'm familiar with a pip-like specification like python >= 3.9 which would pass with 3.9.1.

Homebrew always tries to ship the latest stable version of everything. We also test all packages we update with all current dependencies. This way we know something is working for sure.

This also means that everyone needs to be using the version we tested with to ensure that it works as well as in our tests.

**Reply by jraman** [reply]
CLT is coming up as already installed:

```
$ xcode-select --install
xcode-select: error: command line tools are already installed, use "Software Update" to install updates
```

**Reply by SMillerDev** [reply]
What is the full output of `brew config`? As well as the output of `brew doctor`

**Reply by jraman** [reply]
brew config: https://gist.github.com/jraman/de4b50de24b156193ce25d1feef97b4f

brew doctor: https://gist.github.com/jraman/3d41298cc96c7343f7cceaa97b78e7e9

Thanks

**Reply by SMillerDev** [reply]
Okay, brew doctor revealed the real cause of your issue you're using an unsupported macOS version. That means my previous assumption was wrong and your CLT is likely fine.

That also means that indeed everything will build from source and take hours.

**Reply by jraman** [reply]
Thanks, @SMillerDev, for looking into this.

10.13.6 (High Sierra) is what I have running on my computer.  It's last security update was on November 12. 2020 or about four months ago.  There were updates to later macos versions, but not to 10.13 after that which implies that 10.13 is no longer supported by Apple.

What this means to me (and others in my position I suppose) is either
 1. Keep the OS at 10.13 and incur a high cost of updating or installing software using `brew`.  Because the packages will be built from source, this could take hours.  Also, things could break (for example, `ffmpeg` was broken for me -- it was looking for `libdav1d.4.dylib` which did not exist -- there was only `libdav1d.5.dylib`).  OR
 2. Update the OS and using `brew` should be a breeze -- an OS update may appear to be easy, but is not for me. OR
 3. Don't use `brew` -- try and find some other solution.

None of the above are appealing to me -- maybe it's a Hobson's choice for <span>#</span>2?  Just letting you all know in case someone can do something about it.

In any case, thank you for `brew`.  It is and has been awesome.

**Comment by jraman** [comment]
Summary:
 * `brew install rsync` was done on macos 10.13 which is no longer supported.
 * Installing (or upgrading) on unsupported macos versions results in the packages being built from source.
    - This can take a long time (hours).
    - Many packages in the dependency hierarchy also gets rebuilt from source.
    - The build may not proceed smoothly and could more than likely end up breaking things -- any of the dependent packages may fail to build and this negatively affects the likelihood of the original install succeeding).
 * Best bet is to update your OS and then get back to using `brew`.

**Comment by reedkotler** [comment]
I'm having the same problem. Pip3 will not work for me now either.

reedkotler@Reeds-Mac-mini ~ % brew config
HOMEBREW_VERSION: 3.0.5
ORIGIN: https://github.com/Homebrew/brew
HEAD: 8b677c0ddd5f96b5521c1b353a9ca07dfe57b80e
Last commit: 2 days ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 3db1b5791315b2489557503971aae2d97024cab8
Core tap last commit: 33 minutes ago
Core tap branch: master
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_REPOSITORY: /opt/homebrew
HOMEBREW_CELLAR: /opt/homebrew/Cellar
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.3 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit westmere
Clang: 12.0 build 1200
Git: 2.24.3 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.64.1 => /usr/bin/curl
macOS: 11.2.3-x86_64
CLT: 12.4.0.0.1.1610135815
Xcode: 12.4
Rosetta 2: true
reedkotler@Reeds-Mac-mini ~ % brew postinstall python@3.9
==> Postinstalling python@3.9
==> /opt/homebrew/Cellar/python@3.9/3.9.2_1/bin/python3 -m ensurepip
Last 15 lines from /Users/reedkotler/Library/Logs/Homebrew/python@3.9/post_install.01.python3:
  File "/opt/homebrew/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/opt/homebrew/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/opt/homebrew/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/lib/python3.9/ensurepip/__main__.py", line 5, in <module>
    sys.exit(ensurepip._main())
  File "/opt/homebrew/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/lib/python3.9/ensurepip/__init__.py", line 210, in _main
    return _bootstrap(
  File "/opt/homebrew/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/lib/python3.9/ensurepip/__init__.py", line 129, in _bootstrap
    return _run_pip(args + [p[0] for p in _PROJECTS], additional_paths)
  File "/opt/homebrew/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/lib/python3.9/ensurepip/__init__.py", line 38, in _run_pip
    return subprocess.run([sys.executable, "-c", code], check=True).returncode
  File "/opt/homebrew/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/lib/python3.9/subprocess.py", line 528, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/opt/homebrew/opt/python@3.9/bin/python3.9', '-c', '\nimport runpy\nimport sys\nsys.path = [\'/private/tmp/tmpd33vl09a/setuptools-49.2.1-py3-none-any.whl\', \'/private/tmp/tmpd33vl09a/pip-20.2.3-py2.py3-none-any.whl\'] + sys.path\nsys.argv[1:] = [\'install\', \'--no-cache-dir\', \'--no-index\', \'--find-links\', \'/private/tmp/tmpd33vl09a\', \'setuptools\', \'pip\']\nrunpy.run_module("pip", run_name="__main__", alter_sys=True)\n']' died with <Signals.SIGKILL: 9>.
Warning: The post-install step did not complete successfully
You can try again using:
  brew postinstall python@3.9
reedkotler@Reeds-Mac-mini ~ %
